### ISSUE TITLE
Expose simplify-path

### ISSUE THREAD

**Issue by Rubonnek** [godotengine/godot-proposals#3192]
### Describe the project you are working on

An `EditorImportPlugin` extension.

### Describe the problem or limitation you are having in your project

The data being managed in the import plugin may contains references to other files using relative non-canonical paths which may live outside of `res://`, and `user://`, in which case the import plugin may, or may not, process the referenced files. 

Having a function such as `String.get_canonical_path` that resolves paths similar to the `realpath` function in the GNU C Library could be useful for developing `EditorImportPlugin` extensions since it will allow users to tell whether files live within the project or not, and whether they want to process those files depending on the context where they are used.

Also, as stumbled upon in:

- https://github.com/godotengine/godot/pull/52027

The definitions of what an absolute, relative or canonical path are not clear straight from the engine documentation. These will be updated as part of this proposal.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The `String.get_canonical_path` function will resolve paths when possible, including the cases when `res://` or `user://` are included in the path. In these two cases, these engine-specific references will be replaced with the actual path as consumed by the OS where Godot is running on. This way, `String.get_canonical_path` will be able to resolve paths outside of `res://` and `user://`, and users can use his information to tell whether they want to proceed processing those paths or not, depending on their use case.

The `String.get_canonical_path` function will not check for the existence of files/folders. It will simply create a new cleaned the string that has no references to a parent directory.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Consider the following example implementation in GDScript:

```gdscript
extends Node2D

func _ready():
	var test_strings_array : PoolStringArray = [
			"res:///test/",
			"res://///",
			"res://./",
			"res://./../",
			"res:/",
			"//./",
			"/../",
			"./././",
			"C:\\Users\\Test\\..\\",
			"C:\\..",
			"C:\\\\",
			"test/.../.././",
			"../test",
			"test/..",
			"user:///test/",
			"user:///test/..",
			"user://///",
			"user://./",
			"user://./../",
			"user:/",
			]

	print("Testing get_canonical_path:")
	for test_string in test_strings_array:
		print("\t\"%s\" == \"%s\"" % [ test_string, get_canonical_path(test_string) ])
	
	get_tree().quit()


func get_canonical_path(p_path : String) -> String:
	# Replace Godot-specific references
	p_path = ProjectSettings.globalize_path(p_path)

	# Use forward slash as the path separator
	var replacement : Dictionary = {
		'\\': '/',
		'//': '/'
	}
	for key in replacement:
		p_path = p_path.replace(key, replacement[key])

	# Check if the path begins with a "/" -- we'll use this later.
	var did_path_begin_with_forward_slash : bool = false
	if p_path.begins_with("/"):
		did_path_begin_with_forward_slash = true

	# Split the path so it's easier to manipulate
	var split_path : PoolStringArray = p_path.split("/")

	# If the OS is Windows (or UWP), the path may contain the DOS drive letter.
	var os_name : String = OS.get_name()
	var dos_drive_letter : String = ""
	if os_name == "Windows" or os_name == "UWP":
		dos_drive_letter = split_path[0]
		split_path.remove(0)

	# First pass: Dump all empty entries (i.e. all the "/" characters)
	var remove_index_queue : Array = []
	for index in range(0, split_path.size()):
		# Debug:
		#print(index, ": ", split_path[index], " : ", split_path[index].empty())
		if split_path[index].empty():
			remove_index_queue.push_back(index)
	for _index in range(0, remove_index_queue.size()):
		var popping_index : int = remove_index_queue.pop_back()
		# Debug
		#print("Popping: ", popping_index)
		split_path.remove(popping_index)

	# Second pass: Dump all entries that include the current directory only "i.e. '.'
	for index in range(0, split_path.size()):
		if split_path[index] == ".":
			remove_index_queue.push_back(index)
	for _index in range(0, remove_index_queue.size()):
		var popping_index : int = remove_index_queue.pop_back()
		split_path.remove(popping_index)

	# Third pass: Create upper directory chains and drop them accordingly
	var skip_index_count : int = 0
	for index in range(0, split_path.size()):
		if skip_index_count > 0:
			skip_index_count -= 1
			continue
		if split_path[index] == "..":
			# Queue drop the current path entry
			remove_index_queue.push_back(index)
			# Queue drop next path entries as long as they only refer to the
			# upper directory
			var next_index : int = index + 1
			while true:
				if next_index < split_path.size() and split_path[next_index] == "..":
					skip_index_count += 1
					remove_index_queue.push_back(next_index)
					next_index += 1
				else:
					break
			# At this point we've build a chain of indexes that contain ".." We
			# can use the difference between next index and the current index
			# values to determine how many upper directories we need to remove.
			for parent_relative_index in range(1, next_index - index + 1):
				if (index - parent_relative_index) < 0:
					# Do not try to remove paths that don't exist -- this is
					# how realpath behaves in Linux
					break
				remove_index_queue.push_back(index - parent_relative_index)

	# Sort the remove index queue before we start dropping directories
	remove_index_queue.sort()

	# Drop directories
	for _index in range(0, remove_index_queue.size()):
		var popping_index : int = remove_index_queue.pop_back()
		split_path.remove(popping_index)

	# Fourth pass: reconstruct the final path
	p_path = ""
	for entry in split_path:
		if did_path_begin_with_forward_slash:
			p_path += "/" + entry
		else:
			p_path += entry
			# Toggle the flag to keep prepending forward slashes from now on
			did_path_begin_with_forward_slash = true

	# Fifth pass: (Windows & UWP only) - Reattach the DOS drive letter
	if not dos_drive_letter.empty():
		p_path = dos_drive_letter + "/" + p_path

	return p_path
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This function could be commonly used in `EditorImportPlugin` extensions.

It can be worked around, but not with just a few lines. See the script above.

### Is there a reason why this should be core and not an add-on in the asset library?

Given that it could be commonly used in `EditorImportPlugin` extensions, it makes sense to add it to core since it will facilitate the development of those extensions.

**Comment by Rubonnek** [comment]
Welp, `String.simplify_path` does the same but is not exposed.

**Comment by Calinou** [comment]
Note: Removing the `topic:editor` label because this is a core feature, not an editor-only feature.

**Comment by Rubonnek** [comment]
Closing in favor of keeping the conversation within a set of simple pull requests I'm about to open. I don't see a need to refer back to this proposal since the necessary code is already within the engine.
